numpy/core/einsumfunc.py

```diff
@@ -256,6 +256,108 @@ def _greedy_path(input_sets, output_set, idx_dict, memory_limit):
     return path
 
 
+def _can_blas(inputs, result, idx_removed):
+    """
+    Checks if we can use BLAS call and its beneficial to do so.
+
+    Parameters
+    ----------
+    inputs : list of str
+        Specifies the subscripts for summation.
+    result : str
+        Resulting summation.
+    idx_removed : set
+        Indices that are removed in the summation
+
+
+    Returns
+    -------
+    type : bool
+        Returns true if BLAS should and can be used, else False
+
+    Notes
+    -----
+    If the operations is BLAS level 1 or 2 and is not already aligned
+    we default back to einsum as the memory movement to copy is more
+    costly than the operation itself.
+
+
+    Examples
+    --------
+
+    # Standard GEMM operation
+    >>> _can_blas(['ij', 'jk'], 'ik', set('j'))
+    'GEMM'
+
+    # Can use the standard BLAS, but requires odd data movement
+    >>> _can_blas(['ijj', 'jk'], 'ik', set('j'))
+    False
+
+    # DDOT where the memory is not aligned
+    >>> _can_blas(['ijk', 'ikj'], '', set('ijk'))
+    False
+
+    """
+
+    # Gotta remove indices
+    if len(idx_removed) == 0:
+        return False
+
+    # Can only do two
+    if len(inputs) != 2:
+        return False
+
+    # Make sure there is overlap
+    if len(set(inputs[0]) & set(inputs[1])) == 0:
+        return False
+
+    # Build a few temporaries
+    sets = [set(x) for x in inputs]
+    keep_left = sets[0] - idx_removed
+    keep_right = sets[1] - idx_removed
+    input_left = inputs[0]
+    input_right = inputs[1]
+    rs = len(idx_removed)
+
+    if any(len(l) != len(s) for l, s in zip(inputs, sets)):
+        return False
+
+    # Cannot handle partial inner
+    if len(keep_left & keep_right):
+        return False
+
+    # DDOT
+    elif inputs[0] == inputs[1]:
+        return True
+
+    # DDOT doesnt make sense if you have to tranpose
+    elif sets[0] == sets[1]:
+        return False
+
+    # GEMM or GEMV no transpose
+    elif input_left[-rs:] == input_right[:rs]:
+        return True
+
+    # GEMM or GEMV transpose both
+    elif input_left[:rs] == input_right[-rs:]:
+        return True
+
+    # GEMM or GEMV transpose right
+    elif input_left[-rs:] == input_right[-rs:]:
+        return True
+
+    # GEMM or GEMV tranpose left
+    elif input_left[:rs] == input_right[:rs]:
+        return True
+
+    # Einsum is faster than vectordot if we have to copy
+    elif (len(keep_left) == 0) or (len(keep_right) == 0):
+        return False
+
+    # Conventional matrix tensordot
+    else:
```

Comment:
Thanks for the comments, changes have been made based off them. I would be very open to cleaning up the logic if you have any explicit ideas.